File: 2. Add Two Numbers/Add-Two-Numbers.py
# link : https://leetcode.com/problems/add-two-numbers/submissions/
# 
# Time complexity : O(N + M), where N and M are size of the gven linked lists
# Space cimplexity : O(N), where N is size of linkedList (the returned value)


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        top, bottom, head, carry = l1, l2, ListNode(-1), 0
        curr = head
        
        while (top != None or bottom != None):
            summation = top.val if top else 0
            summation += bottom.val if bottom else 0
            summation += carry
            
            digit = summation % 10
            curr.next = ListNode(digit)
            curr = curr.next
            
            carry = summation // 10
            if top : top = top.next
            if bottom : bottom = bottom.next
        
        if carry: 
            curr.next = ListNode(carry)
        
        return head.next

File: 2. Add Two Numbers/Add-Two-Numbers.java
/** 
 * link : https://leetcode.com/problems/add-two-numbers/
 * 
 * Time complexity : O(N + M), where N and M are size of the gven linked lists
 * Space cimplexity : O(N), where N is size of linkedList (the returned value)
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode top = l1, bottom = l2, head = new ListNode(-1), curr = head;
        int carry = 0, sum = 0;
        
        while(top != null || bottom != null){
            sum = top == null ? 0 : top.val;
            sum += (bottom == null ? 0 : bottom.val) + carry;            
            int digit = sum % 10;

            curr.next = new ListNode(digit);
            curr = curr.next;
            
            carry = sum / 10;
            top = top != null ? top.next : top;
            bottom = bottom != null ? bottom.next : bottom;
        }
        
        if(carry > 0){
            curr.next = new ListNode(carry);
        }
        
        return head.next;
    }
}